# L1: Information Retrieval

In this lab you will apply basic techniques from information retrieval to implement the core of a minimalistic search engine. The data for this lab consists of a collection of app descriptions scraped from the [Google Play Store](https://play.google.com/store/apps?hl=en). From this collection, your search engine should retrieve those apps whose descriptions best match a given query under the vector space model.

## Data set

The app descriptions come in the form of a compressed [JSON](https://en.wikipedia.org/wiki/JSON) file. Start by loading this file into a [Pandas](https://pandas.pydata.org) [DataFrame](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe).

In [198]:
import bz2
import pandas as pd

with bz2.open("app-descriptions.json.bz2") as source:
    df = pd.read_json(source)

In Pandas, a DataFrame is a table with indexed rows and labelled columns of potentially different types. Data in a DataFrame can be accessed in various ways, including by row and by column. To give an example, the code in the next cell shows rows 200–204:

In [199]:
df[200:206]

description  \
200  Introducing the best Brick Breaker game that e...   
201  Classic Brick Game!\n\nBrick Classic is a popu...   
202  Bricks Breaker - Glow Balls is a addictive and...   
203  How to play\n- The ball flies to wherever you ...   
204  Fight brave soldiers from around the globe on ...   
205  The Empire has fallen, and the Age of Great Me...   

                               name  
200  Brick Breaker Star: Space King  
201      Brick Classic - Brick Game  
202     Bricks Breaker - Glow Balls  
203            Bricks Breaker Quest  
204             Brothers in Arms® 3  
205       Brown Dust - Tactical RPG

As you can see, there are two labelled columns: `name` (the name of the app) and `description` (a textual description). The code in the next cell shows how to acess fields from the description column.

In [200]:
df["description"][200:206]

200    Introducing the best Brick Breaker game that e...
201    Classic Brick Game!\n\nBrick Classic is a popu...
202    Bricks Breaker - Glow Balls is a addictive and...
203    How to play\n- The ball flies to wherever you ...
204    Fight brave soldiers from around the globe on ...
205    The Empire has fallen, and the Age of Great Me...
Name: description, dtype: object

## Problem 1: Preprocessing

Your first task is to implement a preprocessor for your search engine. In the vector space model, *preprocessing* refers to any kind of transformation that is applied before a text is vectorized. Here you can restrict yourself to a very simple preprocessing: tokenization, stop word removal, and lemmatization.

To implement your preprocessor, you can use [spaCy](https://spacy.io). Make sure that you read the [Linguistic annotations](https://spacy.io/usage/spacy-101#annotations) section of the spaCy&nbsp;101; that section contains all the information that you need for this problem (and more).

Implement your preprocessor by completing the skeleton code in the next cell, adding additional code as you feel necessary.

In [201]:
def preprocess(text):
    import spacy
    nlp = spacy.load("en_core_web_sm", disable=["tagger","parser"])
    tokenlist = []
    for token in nlp(text):
        if token.is_alpha  and token.is_stop is False:
            if token.lemma_.isalpha():
                tokenlist.append(token.lemma_)      
    return tokenlist

Your implementation should conform to the following specification:

<strong>preprocess</strong> (<em>text</em>)

> Preprocesses given text by tokenizing it, removing any stop words, replacing each remaining token with its lemma (base form), and discarding all lemmas that contain non-alphabetical characters. Returns the list of remaining lemmas (represented as strings).

**Tip:** To speed up the preprocessing, you can disable loading those spaCy components that you do not need, such as the part-of-speech tagger, parser, and named entity recognizer. See [here](https://spacy.io/usage/processing-pipelines#disabling) for more information about this.

Test your implementation by running the following cell:

In [202]:
preprocess("Apple is looking at buying U.K. startup for $1 billion")

['Apple', 'look', 'buy', 'startup', 'billion']

This should give the following output:


## Problem 2: Vectorizing

Your next task is to vectorize the data – and more specifically, to map each app description to a tf–idf vector. For this you can use the [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) class from [scikit-learn](https://scikit-learn.org/stable/). Make sure to specify your preprocessor from the previous problem as the `tokenizer` &ndash; not the `preprocessor`! &ndash; for the vectorizer.

In [203]:
#import numpy as np

# TODO: Replace the next line with your own code.
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer(tokenizer=preprocess)
X = vectorizer.fit_transform(df["description"])

#X = np.zeros((len(df), 1))

Test your implementation by running the following cell:

In [206]:
print(X.shape)

(1614, 20651)


This should show the dimensions of the matrix `X` to be 1614 × 20651.

## Problem 3: Finding terms with low/high idf

Recall that the inverse document frequency (idf) of a term is the lower the more documents from a given collection the term appears in. To get a better understanding for this concept, your next task is to write code to find out which terms have the lowest/highest idf with respect to the app descriptions.

Start by sorting the terms in increasing order of idf, breaking ties by falling back on alphabetic order.

In [63]:
# TODO: Replace the next line with your own code.
#terms = []
idfs = vectorizer.idf_
names = vectorizer.get_feature_names()
terms_unsorted = dict(zip(names,idfs))
terms = sorted(terms_unsorted.items(), key= lambda x:x[1] ,reverse=True)

Now, print out the 10 terms at the top/end of the sorted list:

In [64]:
print(terms[:10])

[('aa', 5.543294782270004), ('abcs', 5.543294782270004), ('abel', 5.543294782270004), ('abide', 5.543294782270004), ('aboard', 5.543294782270004), ('abraham', 5.543294782270004), ('abroad', 5.543294782270004), ('absolute', 5.543294782270004), ('absorber', 5.543294782270004), ('abundant', 5.543294782270004)]


## Problem 4: Retrieving

To complete the search engine, your last task is to write a function that returns the most relevant app descriptions for a given query. An easy way to do solve this task is to use scikit-learn&rsquo;s [NearestNeighbors](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html) class. That class implements unsupervised nearest neighbours learning, and allows you to easily find a predefined number of app descriptions whose vector representations are closest in distance to the query vector.

In [51]:
from sklearn.neighbors import NearestNeighbors
neigh = NearestNeighbors().fit(X)
def search(query):
    # TODO: Replace the next line with your own code.
    query = vectorizer.transform([query])
    i = neigh.kneighbors(query, 10, return_distance=False)[0]
    return df.iloc[i]

Your implementation should conform to the following specification:

<strong>search</strong> (<em>query</em>)

> Returns the 10 app descriptions most similar (in terms of cosine similarity) to the given query as a Pandas DataFrame.

Test your implementation by running the following cell:

In [54]:
search("dodge trains")

description  \
1301  DASH as fast as you can! \nDODGE the oncoming ...   
1300  Subway princess runner, Bus run, forest rush w...   
1428  Master and manage the chaos of international r...   
998   2M+ Downloads All Over The World!\n\n* IGN Nom...   
1394  All aboard for an adventure around the world!\...   
1429  🚂 BabyMagica "Train for Animals" is a educatio...   
1168  Are you ready for a thrilling ride?\n\nRush th...   
1077  Polar Flow is a sports, fitness, and activity ...   
1286  Track your fitness with Strava activity tracke...   
1465  Warning! Virus invasion! Destroy them with you...   

                                           name  
1301                             Subway Surfers  
1300                     Subway Princess Runner  
1428                      Train Conductor World  
998              No Humanity - The Hardest Game  
1394                                 Tiny Rails  
1429        Train for Animals - BabyMagica free  
1168                                       Rush  
1077                Polar Flow – Sync & Analyze  
1286  Strava: Track Running, Cycling & Swimming  
1465            Virus War - Space Shooting Game

The top hit in the list should be *Subway Surfers*.

## Problem 5: Keyword extraction

A simple method for extracting salient keywords from a document is to pick the $k$ terms with the highest tf–idf value. Your last task in this lab is to implement this method. More specifically, we ask you to implement a function `keywords` that extracts keywords from a given text.

In [137]:
#import pandas as pd
#from sklearn.feature_extraction.text import TfidfTransformer
#from sklearn.feature_extraction.text import CountVectorizer
## this is a very toy example, do not try this at home unless you want to understand the usage differences
#text = df["description"][1428] 
#docs=[text]

["Master and manage the chaos of international railway traffic as the ultimate railroad tycoon. Build the rail network of your dreams; lay rails and solve the railroad puzzle with branching and forking roads at every turn. Become the richest manager and pick your path, do you optimise to the micro level, planning routes and managing the timetable, or sit idle letting your business keep earning while you sleep! \n\nGet in the driver's seat and take passengers to their destinations, dropping them at stations, and haul goods to ports and factories. Control and conduct the trains, marshalling them through tunnels, around obstacles and over mountains in this thrilling, fast-paced action arcade videogame. Connect your express trains across the railyard at breakneck speed. Try to avoid crashing, but I bet you can’t. You’ll need snappy strategy to master the chaos! Be on high-alert for explosive crashes, near-misses and split second situations.\n\nRing the bell and toot your horn while playing

In [138]:
##instantiate CountVectorizer()
#v=CountVectorizer(tokenizer=preprocess)
 
## this steps generates word counts for the words in your docs
#word_count_vector=cv.fit_transform(docs)

In [139]:
#word_count_vector.shape

(1, 107)

In [140]:
#tfidf_transformer=TfidfTransformer(smooth_idf=True,use_idf=True)
#tfidf_transformer.fit(word_count_vector)

TfidfTransformer(norm='l2', smooth_idf=True, sublinear_tf=False, use_idf=True)

In [141]:
## print idf values
#df_idf = pd.DataFrame(tfidf_transformer.idf_, index=cv.get_feature_names(),columns=["idf_weights"])
 
## sort ascending
#df_idf = df_idf.sort_values(by=['idf_weights'])

In [142]:
## count matrix
#count_vector=cv.transform(docs)
 
## tf-idf scores
#tf_idf_vector=tfidf_transformer.transform(count_vector)

In [143]:
#feature_names = cv.get_feature_names()
 
##get tfidf vector for first document
#first_document_vector=tf_idf_vector[0]
 
##print the scores
#df = pd.DataFrame(first_document_vector.T.todense(), index=feature_names, columns=["tfidf"])
#df.sort_values(by=["tfidf"],ascending=False)[:10]

tfidf
train     0.518875
pick      0.148250
manage    0.148250
let       0.148250
chaos     0.148250
rail      0.148250
railroad  0.148250
railway   0.148250
master    0.148250
crash     0.148250

In [212]:
def keywords(text, n=10):
    from sklearn.feature_extraction.text import TfidfVectorizer
    tfidf_vectorizer=TfidfVectorizer(tokenizer=preprocess)
    tfidf_vectorizer_vectors=tfidf_vectorizer.fit_transform([text])
    first_vector_tfidfvectorizer=tfidf_vectorizer_vectors[0]
    df = pd.DataFrame(first_vector_tfidfvectorizer.T.todense(), index=tfidf_vectorizer.get_feature_names(), columns=["tfidf"])
    df_tfidf = df.sort_values(by=["tfidf"],ascending=False)[:n]
    df_tfidf_list = []
    for row in df_tfidf.index:
        df_tfidf_list.append(row)
    return df_tfidf_list        

Your implementation should conform to the following specification:

<strong>keywords</strong> (<em>text</em>, <em>n</em> = 10)

> Returns a list with the $n$ (default value: 10) most salient keywords from the specified text, as measured by their tf–idf value relative to the collection of app descriptions.

Test your implementation by running the following cell:

In [213]:
print(keywords(df["description"][1428]))

['train', 'pick', 'manage', 'let', 'chaos', 'rail', 'railroad', 'railway', 'master', 'crash']


This should give the following output:

<div class="alert alert-info">
    Please read the General information section on the lab web page before submitting this notebook!
</div>